stonesoup/dataassociator/general.py

```diff
@@ -0,0 +1,179 @@
+from abc import abstractmethod
+from operator import attrgetter
+from datetime import datetime
+from statistics import mean
+
+from stonesoup.dataassociator.base import Associator
+from stonesoup.types.association import AssociationSet
+from typing import Set, List, Union, Tuple, Dict, Optional
+import numpy as np
+from stonesoup.base import Property, Base
+from stonesoup.measures import GenericMeasure, Measure
+from stonesoup.dataassociator._assignment import assign2D
+from stonesoup.types.association import Association
+from stonesoup.types.track import Track
+from stonesoup.types.state import State
+
+
+class GeneralAssociationGate(Base):
+
+    @abstractmethod
+    def __call__(self, item1, item2) -> bool:
+        raise NotImplementedError
+
+
+class MeasureThresholdGate(GeneralAssociationGate):
+
+    minimise_measure: bool = Property(default=True)
+    association_threshold: float = Property()
+    measure: GenericMeasure = Property()
+
+    def __call__(self, item1, item2) -> bool:
+        distance_measure = self.measure(item1, item2)
+        if self.minimise_measure:
+            return distance_measure <= self.association_threshold
+        else:  # maximise measure
+            return self.association_threshold <= distance_measure
+
+
+class RecentTrackMeasure(GenericMeasure):
+    state_measure: Measure = Property()
+    max_timestep_to_look_back: int = Property(default=10)
+
+    def __call__(self, track1: Track, track2: Track) -> Optional[float]:
+
+        track_1_dict: Dict[datetime, State] = \
+            {state.timestamp: state for i, state in enumerate(reversed(track1.states))
+             if i < self.max_timestep_to_look_back}
+
+        track_2_dict = {state.timestamp: state for i, state in enumerate(reversed(track2.states)) if
+                        i < self.max_timestep_to_look_back}
+
+        all_times = set(track_1_dict.keys()) | set(track_2_dict.keys())
+
+        measures = []
+        for time in all_times:
+            state1 = track_1_dict.get(time)
+            state2 = track_2_dict.get(time)
+
+            if state1 is not None and state2 is not None:
+                measures.append(self.state_measure(state1, state2))
+
+        if len(measures) != 0:
+            return mean(measures)
+        else:
+            return None
+
+
+class OneToOneAssociatorWithGates(Associator):
+
+    gates: List[GeneralAssociationGate] = Property(default=None)
+
+    fail_gate_value: Union[int, float, complex, np.number] = Property(default=None)
+
+    measure: GenericMeasure = Property()
+    association_threshold: float = Property(default=None)
+
+    maximise_measure: bool = Property(default=False)
```

Comment:
All of these  (as well as the class) could do with some documentation.  This is important so the user doesn't have to go through the source code to work out how to use it.  There's plenty of good examples of how to write this if you have a browse around the source files.  If you based this on a paper, a reference to that would be handy as well.